File: srs_tester.py
"""Test and visualization harness for LTE SRS generation."""
from __future__ import annotations

import itertools
from dataclasses import dataclass
from typing import Iterable, List, Tuple

import importlib.util
import numpy as np

from lte_srs import (
    MappingInfo,
    SRSConfig,
    generate_srs,
    normalized_cross_correlation,
)


@dataclass
class SRSResult:
    signal: np.ndarray
    info: MappingInfo


def build_configurations() -> List[SRSConfig]:
    """Create a small suite of diverse SRS configurations for testing."""

    common = dict(bandwidth_config=2, subframe_config=2, b_hop=1, n_ul_rb=50)
    return [
        SRSConfig(
            cell_id=cell,
            group_hopping_enabled=gh,
            sequence_hopping_enabled=sh,
            transmission_comb=comb,
            cyclic_shift=alpha,
            srs_bandwidth=n_b,
            **common,
        )
        for cell, gh, sh, comb, alpha, n_b in [
            (0, True, False, 0, 0.0, 0),
            (1, True, True, 1, np.pi / 4, 1),
            (12, False, True, 0, np.pi / 2, 2),
            (37, False, False, 1, 3 * np.pi / 4, 3),
        ]
    ]


def generate_signals(configs: Iterable[SRSConfig], subframe: int) -> List[SRSResult]:
    results: List[SRSResult] = []
    for cfg in configs:
        sig, info = generate_srs(cfg, subframe)
        results.append(SRSResult(signal=sig, info=info))
    return results


def correlation_matrix(results: List[SRSResult]) -> np.ndarray:
    """Compute the normalized correlation matrix for a set of signals."""

    n = len(results)
    corr = np.zeros((n, n))
    for i, j in itertools.product(range(n), repeat=2):
        corr[i, j] = normalized_cross_correlation(results[i].signal, results[j].signal)
    return corr


def highlight_pairs(corr: np.ndarray, threshold: float = 0.3) -> List[Tuple[int, int, float]]:
    """Return pairs with correlation magnitude above a threshold."""

    pairs: List[Tuple[int, int, float]] = []
    n = corr.shape[0]
    for i in range(n):
        for j in range(i + 1, n):
            if corr[i, j] > threshold:
                pairs.append((i, j, corr[i, j]))
    return pairs


def plot_correlation(corr: np.ndarray, labels: List[str], path: str = "correlation_matrix.png") -> str:
    import matplotlib.pyplot as plt

    fig, ax = plt.subplots(figsize=(6, 5))
    im = ax.imshow(corr, vmin=0, vmax=1, cmap="viridis")
    ax.set_xticks(range(len(labels)))
    ax.set_yticks(range(len(labels)))
    ax.set_xticklabels(labels)
    ax.set_yticklabels(labels)
    plt.setp(ax.get_xticklabels(), rotation=45, ha="right")
    ax.set_title("Normalized SRS Cross-Correlation")
    fig.colorbar(im, ax=ax, label="|corr|")
    fig.tight_layout()
    fig.savefig(path, dpi=150)
    plt.close(fig)
    return path


def has_matplotlib() -> bool:
    return importlib.util.find_spec("matplotlib") is not None


def main() -> None:
    configs = build_configurations()
    subframe = 4
    print(f"Generating SRS for subframe {subframe}\n")
    for cfg in configs:
        print(cfg)

    results = generate_signals(configs, subframe)
    labels = [f"UE{i}" for i in range(len(results))]
    corr = correlation_matrix(results)
    pairs = highlight_pairs(corr)

    print("\nCorrelation matrix (magnitude):")
    with np.printoptions(precision=3, suppress=True):
        print(corr)

    if pairs:
        print("\nHighly correlated pairs (>0.3):")
        for i, j, v in pairs:
            print(f"  {labels[i]} vs {labels[j]}: {v:.3f}")
    else:
        print("\nNo highly correlated pairs detected.")

    if has_matplotlib():
        plot_path = plot_correlation(corr, labels)
        print(f"Correlation heatmap saved to {plot_path}")
    else:
        print("Matplotlib is not available; skipping heatmap rendering.")


if __name__ == "__main__":
    main()

File: matlab/srs_tester.m
function srs_tester()
%SRS_TESTER Generate sample LTE SRS waveforms and visualize correlation.
%   This MATLAB version mirrors the Python srs_tester.py script. It builds
%   a small suite of UE configurations, generates SRS symbols, computes the
%   normalized cross-correlation matrix, highlights highly-correlated pairs,
%   and saves a heatmap to correlation_matrix.png.

addpath(fileparts(mfilename('fullpath')));

configs = build_configurations();
subframe = 4;
fprintf('Generating SRS for subframe %d\n\n', subframe);
for idx = 1:numel(configs)
    disp(configs{idx});
end

results = generate_signals(configs, subframe);
labels = arrayfun(@(i) sprintf('UE%d', i - 1), 1:numel(results), 'UniformOutput', false);
corr = correlation_matrix(results);
pairs = highlight_pairs(corr);

fprintf('\nCorrelation matrix (magnitude):\n');
disp(corr);

if isempty(pairs)
    fprintf('\nNo highly correlated pairs detected.\n');
else
    fprintf('\nHighly correlated pairs (>0.3):\n');
    for idx = 1:size(pairs, 1)
        fprintf('  %s vs %s: %.3f\n', labels{pairs(idx, 1)}, labels{pairs(idx, 2)}, pairs(idx, 3));
    end
end

if license('test', 'MAP_Toolbox') || license('test', 'MATLAB')
    plot_path = plot_correlation(corr, labels);
    fprintf('Correlation heatmap saved to %s\n', plot_path);
else
    fprintf('MATLAB graphics not available; skipping heatmap rendering.\n');
end
end

function configs = build_configurations()
    common = {'bandwidth_config', 2, 'subframe_config', 2, 'b_hop', 1, 'n_ul_rb', 50};
    entries = {
        0, true, false, 0, 0.0, 0;
        1, true, true, 1, pi/4, 1;
        12, false, true, 0, pi/2, 2;
        37, false, false, 1, 3*pi/4, 3;
    };
    configs = cell(1, size(entries, 1));
    for idx = 1:size(entries, 1)
        row = entries(idx, :);
        configs{idx} = lte_srs.create_config('cell_id', row{1}, ...
            'group_hopping_enabled', row{2}, ...
            'sequence_hopping_enabled', row{3}, ...
            'transmission_comb', row{4}, ...
            'cyclic_shift', row{5}, ...
            'srs_bandwidth', row{6}, ...
            common{:});
    end
end

function results = generate_signals(configs, subframe)
    results = cell(size(configs));
    for idx = 1:numel(configs)
        [sig, info] = lte_srs.generate_srs(configs{idx}, subframe);
        results{idx} = struct('signal', sig, 'info', info);
    end
end

function corr = correlation_matrix(results)
    n = numel(results);
    corr = zeros(n);
    for i = 1:n
        for j = 1:n
            corr(i, j) = lte_srs.normalized_cross_correlation(results{i}.signal, results{j}.signal);
        end
    end
end

function pairs = highlight_pairs(corr, threshold)
    if nargin < 2
        threshold = 0.3;
    end
    pairs = [];
    n = size(corr, 1);
    for i = 1:n
        for j = i+1:n
            if corr(i, j) > threshold
                pairs = [pairs; i, j, corr(i, j)]; %#ok<AGROW>
            end
        end
    end
end

function plot_path = plot_correlation(corr, labels)
    figure('Visible', 'off');
    imagesc(corr, [0 1]);
    colormap('viridis');
    colorbar;
    set(gca, 'XTick', 1:numel(labels), 'YTick', 1:numel(labels), ...
        'XTickLabel', labels, 'YTickLabel', labels, 'TickDir', 'out');
    xtickangle(45);
    title('Normalized SRS Cross-Correlation');
    axis square;
    plot_path = fullfile(fileparts(mfilename('fullpath')), 'correlation_matrix.png');
    saveas(gcf, plot_path);
    close(gcf);
end
